include/fi_tagged_rx.h

```diff
@@ -0,0 +1,166 @@
+ *
+ * <name>_entry_alloc - gets an entry from freestack. Caller can insert
+ * it into the posted/unexpected list as needed
+ *
+ * <name>_search_post - search for a matching recv entry with the given
+ * tag/addr in posted list
+ * 
+ * <name>_search_unexp - search for a matching recv entry with the
+ * given tag/match/addr in unexpected list
+ */
+#define DECLARE_TAGGED_RX(entrytype, name)				\
+DECLARE_FREESTACK(entrytype, name ## _fs)				\
+struct name {								\
+	struct dlist_entry post_list;					\
+	struct dlist_entry unexp_list;					\
+	struct name ##_fs *fs;						\
+};									\
+									\
+static inline struct name *name ## _create(ssize_t num_rx)		\
+{									\
+	int idx, i;							\
+	entrytype *entry;						\
+	struct name *list;						\
+	struct dlist_entry *item;					\
+	list = calloc(1, sizeof(struct name));				\
+	if (!list)							\
+		return NULL;						\
+									\
+	list->fs = name ## _fs_create(num_rx);				\
+	if (!list->fs) {						\
+		free(list);						\
+		return NULL;						\
+	}								\
+									\
+	dlist_init(&list->post_list);					\
+	dlist_init(&list->unexp_list);					\
+	for (i = 0; i < num_rx; i++) {					\
+		idx = freestack_pop(list->fs);				\
+		assert(idx != FREESTACK_EMPTY);				\
+		entry = &list->fs->buf[idx];				\
+		entry->index = idx;					\
+		dlist_insert_tail(&entry->entry, &list->post_list);	\
+	}								\
+	for (i = 0; i < num_rx; i++) {					\
+		item = list->post_list.next;				\
+		dlist_remove(item);					\
+		entry = container_of(item, entrytype, entry);		\
+		freestack_push(list->fs, entry->index);			\
+	}								\
+	return list;							\
+}									\
+									\
+static inline void name ## _free(struct name *list)			\
+{									\
+	name ## _fs_free(list->fs);					\
+	free(list);					        	\
+}									\
+									\
+static inline entrytype *name ## _entry_alloc(struct name *list)	\
+{									\
+	int idx = freestack_pop(list->fs);				\
+	assert(idx != FREESTACK_EMPTY);					\
+	return &list->fs->buf[idx];					\
+}									\
+									\
+static inline void name ## _entry_free(struct name *list,		\
+				       entrytype *entry)		\
+{									\
+	freestack_push(list->fs, entry->index);				\
+}									\
+									\
+static inline entrytype *name ## _search_post(struct name *list,	\
+					      uint64_t tag,		\
+					      fi_addr_t addr)		\
+{									\
+	entrytype *entry;						\
+	struct dlist_entry *item;					\
+	dlist_foreach(&list->post_list, item) {				\
+		entry = container_of(item, entrytype, entry);		\
+		if ((entry->tmsg.tag & ~entry->tmsg.ignore) ==		\
+		    (tag & ~entry->tmsg.ignore) &&			\
+		    ((entry->tmsg.addr == FI_ADDR_UNSPEC) ||		\
+		     (addr == FI_ADDR_UNSPEC) ||			\
+		     (addr == entry->tmsg.addr)))			\
+			return entry;					\
+	}								\
+	return NULL;							\
+}									\
+									\
+static inline entrytype *name ## _search_unexp(struct name *list,	\
+					       uint64_t tag,		\
+					       uint64_t match,		\
+					       fi_addr_t addr)		\
```

Comment:
In fact, can't we do this operation instead?

(entry->tag | entry->ignore) == (tag | entry->ignore)